### PR TITLE
Adds dendrite/sparse mlp models, modules, and policies, and integrates with garage. 

### PR DESCRIPTION
**Pull request by akashvelu**
This isn't fully cleaned yet, but the main code is there - just opening this PR so you all can take a look

### CODE REVIEW

**Review by vkakerbeck — APPROVED** [review]

**Comment by akashvelu** [comment]
Code has been updated to incorporate flake8 formattin

File: nupic/embodied/algos/sparse_weights_sac.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        if self._use_automatic_entropy_tuning:
+            alpha_loss = self._temperature_objective(log_pi_new_actions,
+                                                     samples_data)
+            self._alpha_optimizer.zero_grad()
+            alpha_loss.backward()
+            self._alpha_optimizer.step()
+
+        return policy_loss, qf1_loss, qf2_loss
+
+    def _evaluate_policy(self, epoch):
```
Comment (lucasosouza) [inline nupic/embodied/algos/sparse_weights_sac.py:88]: Shouldn't this be the default behavior for all algorithms (to evaluate in eval mode), whether they are sparse or not? If that is the case, why it isn't implemented like this in garage?

**Review by lucasosouza — APPROVED** [review]
Left tons of comments and questions. Please go through them at your own will and pace. You can merge nevertheless if it is blocking your progress.

I haven't run it yet, so I might have more comments and questions later. If you are comfortable with it, I can also send some pull requests later if I spot any improvements or bug fixes. 

On the "java-esque" approach to having one class per file, I don't disagree with it, as we discussed. But as the library grows and more contributors are added, that will quickly become untractrable, we will have dozens of .py file under each folder. Please consider at some point consolidating related classes into a single module. In python, each .py can be considered a module, so it is natural to have several functions and classes within each file. That will also make automatic documentation easier (for example when using Sphinx). 

When doing this integration, please also consider subclassing within each module to remove some of the duplicated code. For example, can multiheaded_sparse subclass from sparse_mlp (instead of subclassing from nn.Module), and only have the few required modifications that characterize multiheaded? If we can do that, it will be make the code easier to read and mantain.

File: nupic/embodied/models/dendrite_mlp.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            sparsity=sparsity,
+            kw=kw,
+            relu=relu,
+            output_nonlinearities=output_nonlinearities,
+            dendritic_layer_class=dendritic_layer_class
+        )
+        self._output_dim = output_dim
+
+    def forward(self, x, context):
+        return super().forward(x, context)[0]
```
[the remark this answers is not in the document]
Comment (akashvelu) [reply nupic/embodied/models/dendrite_mlp.py:61]: This is to get the output of the 1st head of the multi-headed output (there's only 1 head since it's a normal MLP)

File: projects/garage_experiments/sparse_mlp_mtsac_mt1_pickplace.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    mt1 = metaworld.MT1('pick-place-v1')
+    mt1_test = metaworld.MT1('pick-place-v1')
+    train_task_sampler = MetaWorldTaskSampler(mt1, 'train',
+                                              lambda env, _: normalize(env))
+    test_task_sampler = MetaWorldTaskSampler(mt1_test, 'train',
+                                             lambda env, _: normalize(env))
+    n_tasks = 50
+    train_envs = train_task_sampler.sample(n_tasks)
+    env = train_envs[0]()
+    test_envs = [env_up() for env_up in test_task_sampler.sample(n_tasks)]
```
[the remark this answers is not in the document]
Comment (akashvelu) [reply projects/garage_experiments/sparse_mlp_mtsac_mt1_pickplace.py:49]: Yep, that's handled in the train code

File: nupic/embodied/models/dendrite_mlp.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    """
+
+    def __init__(self,
+                 input_size,
+                 output_dim,
+                 dim_context,
+                 hidden_sizes=(32, 32),
+                 num_segments=(5, 5),
+                 sparsity=0.5,
+                 kw=False,
```
[the remark this answers is not in the document]
Comment (akashvelu) [reply nupic/embodied/models/dendrite_mlp.py:40]: Fixed

File: nupic/embodied/models/multiheaded_dendrite_mlp.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+from garage.torch import NonLinearity
+from nupic.research.frameworks.dendrites import (
+    AbsoluteMaxGatingDendriticLayer,
+    DendriticAbsoluteMaxGate1d,
+    DendriticGate1d,
+)
+from nupic.torch.modules import KWinners
+
+
+class MultiHeadedDendriticMLP(nn.Module):
```
[the remark this answers is not in the document]
Comment (akashvelu) [reply nupic/embodied/models/multiheaded_dendrite_mlp.py:33]: Added!

File: projects/garage_experiments/sparse_mlp_mtsac_mt1_pickplace.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        num_tasks=1,
+        steps_per_epoch=epoch_cycles,
+        replay_buffer=replay_buffer,
+        min_buffer_size=1500,
+        target_update_tau=5e-3,
+        discount=0.99,
+        buffer_batch_size=1280)
+    if _gpu is not None:
+        set_gpu_mode(True, _gpu)
+    mtsac.to()
```
[the remark this answers is not in the document]
Comment (akashvelu) [reply projects/garage_experiments/sparse_mlp_mtsac_mt1_pickplace.py:108]: This isn't calling to() on a model; it's a function in the algorithm (SAC) class which moves all networks to the device. The device is determined by a variable set by set_gpu_mode.

File: projects/garage_experiments/sparse_mlp_mtsac_mt1_pickplace.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    trainer.setup(algo=mtsac, env=train_envs)
+    trainer.train(n_epochs=epochs, batch_size=batch_size)
+
+
+# pylint: disable=missing-kwoa
```
[the remark this answers is not in the document]
Comment (akashvelu) [reply projects/garage_experiments/sparse_mlp_mtsac_mt1_pickplace.py:113]: Fixed

File: projects/garage_experiments/mtsac_mt10.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+"""MTSAC implementation based on Metaworld. Benchmarked on MT10.
```
[the remark this answers is not in the document]
Comment (akashvelu) [reply projects/garage_experiments/mtsac_mt10.py:2]: Fixed

File: nupic/embodied/q_functions/continuous_dendrite_mlp_q_function.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        self._obs_dim = env_spec.observation_space.flat_dim - self._dim_context
+        self._action_dim = env_spec.action_space.flat_dim
+
+        DendriticMLP.__init__(self,
+                              input_size=self._obs_dim + self._action_dim,
+                              output_dim=1,
+                              dim_context=dim_context,
+                              **kwargs)
+
+    # pylint: disable=arguments-differ
```
[the remark this answers is not in the document]
Comment (akashvelu) [reply nupic/embodied/q_functions/continuous_dendrite_mlp_q_function.py:55]: Fixed

File: nupic/embodied/models/multiheaded_dendrite_mlp.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            if kw:
+                curr_activation = KWinners(n=hidden_sizes[i],
+                                           percent_on=0.05,
+                                           k_inference_factor=1.0,
+                                           boost_strength=0.0,
+                                           boost_strength_factor=0.0)
+            else:
+                curr_activation = nn.ReLU()
+
+            self._layers.append(curr_dend)
```
[the remark this answers is not in the document]
Comment (akashvelu) [reply nupic/embodied/models/multiheaded_dendrite_mlp.py:105]: This is a little tricky (as we found out on Friday) since Dendrite takes in both feedforward input and context, so this is a little cleaner for now I think.

File: nupic/embodied/models/dendrite_mlp.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+#  along with this program.  If not, see http://www.gnu.org/licenses.
+#
+#  http://numenta.org/licenses/
+#
+# ------------------------------------------------------------------------------
+from nupic.embodied.models import MultiHeadedDendriticMLP
+from nupic.research.frameworks.dendrites import AbsoluteMaxGatingDendriticLayer
+
+
+class DendriticMLP(MultiHeadedDendriticMLP):
```
[the remark this answers is not in the document]
Comment (akashvelu) [reply nupic/embodied/models/dendrite_mlp.py:26]: Added to multi-headed dendrite MLP

File: nupic/embodied/algos/sparse_weights_sac.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (akashvelu) [reply nupic/embodied/algos/sparse_weights_sac.py:88]: I agree; I couldn't find it in their code so I added it. They wrap all environment collection steps with a torch.no_grad(), though.
